[177185] Wrapper library for zip

aNobuhisaMasaki opened the pull request:
https://acsmine.tok.access-company.com/redmine/issues/177185

Abstract:
Implemented a wrapper module for `zip` shell command.

Purpose:
To add feature to generate password-encrypted ZIP file from a *gear*.
I chose to use shell command because it is the only cost-effective way which can suffice such a feature.

Limitations:
Along with newly utilizing a shell command, it is aided with limited permission to the local file systems.
In detail,
I added an interface `AntikytheraCore.TmpdirTracker.get/1` to get the path to currently used `tmpdir`.
After applying this interface, it is ensured:
- that it can be called from only within `Tmpdir.make/2` call,
- that the path for input file/directory is under `tmpdir`,
- that the path for resulting archive is under `tmpdir`.

I appreciate your suggestions.

aNobuhisaMasaki commented on test/lib/util/zip_test.exs:
```diff
@@ -0,0 +1,109 @@
+          assert Zip.zip(@context, zip_path, tmpdir <> src_path, [encryption: true, password: "password"]) == {:ok, zip_path}
+        end)
+      end)
+    end
```
ここまでの2件のテストはblackbox testであり、実際に`zip`コマンドを実行します。
`System.cmd/3`に対するmockが効かなかったためで、mockできる方法があればそちらに置き換えたいと思います。
具体的には、
```elixir
:meck.expect(System, :cmd, fn(_, _, _) -> {"ok", 0} end)
```
としてもこの関数には制御が移らずに、本来の`System.cmd/3`が実行されます。

aNobuhisaMasaki commented on test/lib/util/zip_test.exs:
```diff
@@ -0,0 +1,109 @@
+        end)
+        assert {:error, {:shell_runtime_error, _}} = Zip.zip(@context, zip_path, src_path)
+      end)
+    end
```
この1件のテストもblackbox testです。
入力元のファイルを作成せずに`File.exists?/1`をmockして矛盾を発生させています。

aNobuhisaMasaki commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,100 @@
+      :ok           <- validate_within_tmpdir(zip_path, tmpdir),
+      :ok           <- validate_within_tmpdir(src_path, tmpdir),
+      :ok           <- validate_path_exists(zip_path, false),
+      :ok           <- validate_path_exists(src_path, true),
```
Probably I put toomuch emphasis on the case when `zip_path` is a *file* path...
Actually, it needs another way to treat with `zip_path` given as a *directory* path:
<table>
  <thead>
    <tr>
      <th>`zip_path` is a ...</th>
      <th>Result</th>
      <th>Conclution</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Existent file</td>
      <td>Error by `zip/3`</td>
      <td>Consistent</td>
    </tr>
      <td>Non-existent file</td>
      <td>OK by `zip/3`</td>
      <td>Consistent</td>
    </tr>
      <td>Existent directory</td>
      <td>OK by shell / Bad args for `zip/3`</td>
      <td>Inconsistent</td>
    </tr>
      <td>Non-existent directory</td>
      <td>Error by shell / Good args for `zip/3`</td>
      <td>Inconsistent</td>
    </tr>
  </tbody>
</table>

aNobuhisaMasaki replied:
Maybe it should accept *`Existent file`*; there is the *update* behavior in `zip` command.
Then, all it rejects is *`Non-existent directory`* as for `zip_path`.

aNobuhisaMasaki replied:
Actually, there is a way to crack the update behavior: update non-zip file by zip command.
But I wonder it can be validated in advance.
For now, it will be delegated to `zip` command, because it is too hard for a *gear* to look into detailed file format.

aNobuhisaMasaki replied:
I will further check edge cases in [`:zip.zip/3`](http://erlang.org/doc/man/zip.html#zip-3)...

aNobuhisaMasaki replied:
Maybe there are major changes:
- Behavior of `:zip.zip/3`:
  - when `zip_path` is a directory:
    - `mkdir -p` unless it exists
    - Returns `{:error, :eisdir}` even if the directory is created.
  - when `zip_path` is a file:
    - `mkdir -p` unless the directory exists
    - `touch` unless the file exists
    - Returns `{:error, :enoent}` unless `src_path` exists, even if the file is created.

aNobuhisaMasaki replied:
I decided to avoid some features which result in empty residues:
- ~when `zip_path` is a directory:~
  - ~`mkdir -p` unless it exists acf798e~
  - ~Returns `{:error, :eisdir}` even if the directory is created.~
  - => Invalidated; `zip_path` must be a file! 521977b 9f7681f
- when `zip_path` is a file:
  - `mkdir -p` unless the directory exists acf798e
  - `touch` unless the file exists 43f6c27
  - ~Returns `{:error, :enoent}` unless `src_path` exists, even if the file is created. 43f6c27~
  - => `touch` after checking `src_path` exists. There will not be an empty file. 
3b33d7a e17c625

aNobuhisaMasaki commented:
Let me summarize again.

`Antikythera.Zip`'s features are understood by comparison between Erlang's [`:zip.zip/3`](http://erlang.org/doc/man/zip.html#zip-3).

- Definitions of terms
  - Raw path
    - User input of the path including path traversal (`../`) and current directory (`./`), like:
      - `/a/b/c/../../d`(finally `/a/d`)
  - Detour
    - Sequence of paths made from raw path, like:
      - `/a/b/c/../../d` => `/a/`,`/a/b/`,`/a/b/c/`,(`/a/b/`),(`/a/`),`/a/d`
  - Parsed path
    - The path calculated by removing path traversal (`../`) and current directory (`./`), like:
      - `/a/b/c/../../d` => `/a/d`
  - Direct line
    - Sequence of paths made from parsed path, like:
      - (`/a/b/c/../../d` =>) `/a/d` => `/a/`,`/a/d`
<table>
  <thead>
    <tr>
      <th>Case</th>
      <th>`:zip.zip/3`</th>
      <th></th>
      <th>`Antikythera.Zip.zip/3`</th>
      <th></th>
    </tr>
    <tr>
      <th></th>
      <th>method</th>
      <th>product</th>
      <th>method</th>
      <th>product</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Evaluation of `zip_path`</td>
      <td>Tries `mkdir -p` for every part of raw path.</td>
      <td>Detour into the `zip_path`.</td>
      <td>Tries `mkdir -p` for every part of parsed path.</td>
      <td>Direct line into the `zip_path`.</td>
    </tr>
    <tr>
      <td>Problems of directory `zip_path`, which ends with `/`, `/.` or `/..`</td>
      <td>Tries `mkdir -p` then raises an argument error.</td>
      <td>Unused detour.</td>
      <td>Raises an argument error in advance.</td>
      <td>Nothing.</td>
    </tr>
    <tr>
      <td>Problems of existent file as directory like `file.txt/`</td>
      <td>Evaluates every part of the raw path.</td>
      <td>As long a detour as is valid.</td>
      <td>Evaluates every part of the parsed path.</td>
      <td>As long a direct line as is valid.</td>
    </tr>
    <tr>
      <td>Problems of non-existent directory in `src_path`</td>
      <td>Evaluates every part of the raw path.</td>
      <td>Nothing.</td>
      <td>Evaluates every part of the parsed path (*REMARK*).</td>
      <td>Nothing.</td>
    </tr>
  </tbody>
</table>

- *REMARK*
  - This approach is the best in that it hides the directory structures outside `tmpdir`.
  - There is a possibility that it reflects the directory structures inside `tmpdir`, but omitted for now, because of simplicity and that it is only an exception that there is such a path traversal.

aNobuhisaMasaki commented:
Locally tested using `iex`:
```elixir
$ iex -S mix
iex> context = Antikythera.Test.ConnHelper.make_conn().context
iex> Antikythera.Tmpdir.make(context, fn tmpdir ->
  src = tmpdir <> "/src.txt"
  zip = tmpdir <> "/z.zip"
  File.write!(src, "foo")
  Antikythera.Zip.zip(context, zip, src)
end)
```

aMasakiTakahashi commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,135 @@
+  ) :: R.t(Path.t) do
+    epool_id = extract_epool_id(context_or_epool_id)
+    with(
+      {:ok, tmpdir} <- TmpdirTracker.get(epool_id),
```
If `zip/4` is called without creating `Tmpdir`, this function returns following.
```
{:error,
 {:not_found,
  %{
    state: %AntikytheraCore.TmpdirTracker.State{
      gear_tmp_dir: "/Users/takahashimasaki/src/antikythera/tmp/66687/gear_tmp",
      map: %{}
    }
  }}}
```
`state` is implementation detail, so it should not be returned.
Returing only error reason(ex.{:error, :not_found}) is better.

aNobuhisaMasaki replied:
Removed `state` in 89fa59c .

aMasakiTakahashi commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,135 @@
+      {_,   0}      ->
+        :ok
+      {msg, status} ->
+        {:error, {:shell_runtime_error, %{msg: msg, status: status}}}
```
It would be better that `zip/4` doesn't depend on zip command.
Therefore, return value should not contain `mag` and `status`.

aNobuhisaMasaki replied:
Removed map, then it simply returns `{:error, :shell_runtime_error}`. 6e78235

aMasakiTakahashi commented on test/core/sup_tree_core/tmpdir_tracker_test.exs:
```diff
@@ -0,0 +1,22 @@
+  @context Antikythera.Test.ConnHelper.make_conn().context
+
+  test "TmpdirTracker.get/1 returns tmpdir only within Tmpdir.make/2" do
+    assert %{
```
Following is more simple.
```
Tmpdir.make(@context, fn tmpdir ->
  assert {:ok, tmpdir} == TmpdirTracker.get(@context.executor_pool_id)
end)
```

aNobuhisaMasaki replied:
Fixed in 1bb0d20 .

aMasakiTakahashi commented on test/lib/util/zip_test.exs:
```diff
@@ -0,0 +1,131 @@
+    end
+
+    test "returns error when tmpdir is not found" do
+      :meck.expect(TmpdirTracker, :get, fn _ -> {:error, {:not_found, %{}}} end)
```
This line is not necessary.

aNobuhisaMasaki replied:
Removed in 53f16aa .

aMasakiTakahashi commented on test/lib/util/zip_test.exs:
```diff
@@ -0,0 +1,131 @@
+
+    test "returns error when src is not found" do
+      :meck.expect(TmpdirTracker, :get, fn _ -> {:ok, @tmpdir} end)
+      :meck.expect(File, :exists?, fn @src_path -> false end)
```
This line is not necessary.

aNobuhisaMasaki replied:
Removed in 53f16aa .

aMasakiTakahashi commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,135 @@
+        {:ok,    ["-P", password]}
+      %{encryption: false, password: _}        ->
+        {:error, {:argument_error, map}}
+      _                                        ->
```
In case `encryption` is `true` and `password` is not specified, `Zip.zip/4` should return argument error tuple.
(Currently, it returns `{:error, :shell_runtime_error}`)

aNobuhisaMasaki replied:
Added validations in 3546660 .

aMasakiTakahashi commented:
LGTM, thanks!
Please merge after all other members approved.

aNobuhisaMasaki commented:
@aMasakiTakahashi Thank you for the review!

> all other members

Excuse me, but who are in the scope?
Anyway, I cannot assign reviewers for now.

ttokit commented on test/core/sup_tree_core/tmpdir_tracker_test.exs:
```diff
@@ -0,0 +1,16 @@
+# Copyright(c) 2015-2019 ACCESS CO., LTD. All rights reserved.
+
+defmodule AntikytheraCore.TmpdirTrackerTest do
+  use ExUnit.Case
```
In `sup_tree_core/*_test.exs`, `use Croma.TestCase` is used instead of `use ExUnit.Case`.
For a sense of unity, let's use `Croma.TestCase` here as well.

aNobuhisaMasaki replied:
Replaced with `Croma.TestCase`. 63d3b2f

aNobuhisaMasaki commented:
@aMasakiTakahashi @ttokit 
- Added minor fixes:
  - Fixed regex pattern to accept file name like `/file.` or `/...`: f277b42
  - Line breaking is rejected from file name: 5f6193e
  - Fixed response when `.zip` extension is added: 287f7a6

aMasakiTakahashi commented:
@aNobuhisaMasaki 
>Excuse me, but who are in the scope?
>Anyway, I cannot assign reviewers for now.

@ttokit and @toyoaki-k 
(You need not assign reviewers.)

ttokit commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,143 @@
+    zip_raw_path        :: v[FileName.t],
+    src_raw_path        :: v[String.t],
+    opts                :: v[list(opts)] \\ []
+  ) :: R.t(Path.t) do
```
As it is written as follows in [Antikythera Style Guide](https://hexdocs.pm/antikythera/style_guide.html), please be aware of the style unification of the existing codes.
> 2. Be consistent with surrounding code.

The following format is generally used within Antikythera.
```ex
defun zip(context_or_epool_id :: v[EPoolId.t | Context.t],
          zip_raw_path        :: v[FileName.t],
          src_raw_path        :: v[String.t],
          opts                :: v[list(opts)] \\ []) :: R.t(Path.t) do

aNobuhisaMasaki replied:
Fixed in 2a164a4 .

ttokit commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,143 @@
+      :ok           <- ensure_dir_exists(zip_path, tmpdir),
+      {:ok, args}   <- opts |> Map.new() |> extract_zip_args(),
+      :ok           <- try_zip_cmd(args ++ [zip_path, src_path])
+    ) do
```
For the same reason as https://github.com/access-company/antikythera/pull/44#discussion_r288398820:
```
with {:ok, tmpdir} <- TmpdirTracker.get(epool_id),
     ...,
     ... do
```

However, this comment is optional because I think it is easy to read even in the current writing style.

aNobuhisaMasaki replied:
Fixed in 2a164a4 .

> it is easy to read even in the current writing style.

It could be, but it also depends on how to make decisions about coding style.

ttokit commented on test/core/sup_tree_core/tmpdir_tracker_test.exs:
```diff
@@ -0,0 +1,16 @@
+defmodule AntikytheraCore.TmpdirTrackerTest do
+  use Croma.TestCase
+  alias Antikythera.Tmpdir
+  alias AntikytheraCore.TmpdirTracker
```
This line is no longer needed. (also `ZipTest`)
`Croma.TestCase` automatically alias the module under test.
For more details see [doc](https://hexdocs.pm/croma/Croma.TestCase.html).

aNobuhisaMasaki replied:
Applied alias in bf9eb99 .

ttokit commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,139 @@
+        {:error, {:argument_error, map}}
+      %{encryption: false, password: _}        ->
+        {:error, {:argument_error, map}}
+      _                                        ->
```
These white spaces may occur unnecessary maintenance costs.
When straddling lines, it is not necessary to align `->`.
Please do one of the following.
1. one line
```ex
    case map do
      %{password: ""}                          -> {:error, {:argument_error, map}}
      %{encryption: true,  password: password} -> {:ok,    ["-P", password]      }
      %{encryption: true}                      -> {:error, {:argument_error, map}}
      %{encryption: false, password: _}        -> {:error, {:argument_error, map}}
      _                                        -> {:ok,    []                    }
    end
```
2. multiple line
```ex
    case map do
      %{password: ""} ->
        {:error, {:argument_error, map}}
      %{encryption: true,  password: password} ->
        {:ok, ["-P", password]}
      %{encryption: true} ->
        {:error, {:argument_error, map}}
      %{encryption: false, password: _} ->
        {:error, {:argument_error, map}}
      _ ->
        {:ok, []}
    end
```

aNobuhisaMasaki replied:
Removed indentation in 963c0ea .

ttokit commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,139 @@
+    case System.cmd("zip", args) do
+      {_,   0} ->
+        :ok
+      _        ->
```
Same as https://github.com/access-company/antikythera/pull/44#discussion_r288456661.

aNobuhisaMasaki replied:
Removed indentation in 963c0ea .

ttokit commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,139 @@
+  @typep opts :: {:encryption, boolean} | {:password, String.t}
+
+  defmodule FileName do
+    use Croma.SubtypeOfString, pattern: ~r/^(?!.*\/\.{0,2}\z).*\z/
```
```suggestion
    use Croma.SubtypeOfString, pattern: ~R/^(?!.*\/\.{0,2}\z).*\z/
```

ttokit commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,139 @@
+  defp extract_epool_id(%Context{executor_pool_id: epool_id}), do: epool_id
+  defp extract_epool_id(epool_id),                             do: epool_id
+
+  defunp validate_suffix(path :: v[String.t]) :: :ok | {:error, tuple} do
```
It's a trivial point, but let's change the order of function definition of `validate_suffix/1` and `reject_existing_dir/1`. (Or the order in the `with` expression)

aNobuhisaMasaki replied:
Changed `defunp` order in 0611d0b .

ttokit commented:
LGTM!

aNobuhisaMasaki commented:
@ttokit Thank you for the review!
@toyoaki-k Could you give me some suggestions?

toyoaki-k commented on lib/util/zip.ex:
```diff
@@ -0,0 +1,139 @@
+      case File.mkdir_p(dirname) do
+        :ok ->
+          :ok
+        {:error, :eexist} ->
```
@aNobuhisaMasaki 
Isn't `:eexist` typo?

aNobuhisaMasaki replied:
Seems OK; how can I reproduce `File.mkdir_p(dirname) == :eexist`?
```elixir
iex(12)> File.touch("a")   
:ok
iex(13)> File.mkdir_p("a") 
{:error, :eexist}
```

toyoaki-k replied:
Sorry! I misunderstood.
Seems OK.

toyoaki-k commented:
LGTM!

aNobuhisaMasaki commented:
@toyoaki-k Thank you for the review!

@aMasakiTakahashi @ttokit 
Proceeding to merge...
